pkg/util/config/config.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/spf13/viper"
+)
+
+type config interface {
+	Set(key string)
+	Unset(key string)
+	IsSet(key string) bool
+	Normalize()
+}
+
+// Unmarshal unmarshals the config from the viper into the structure.
+// First, it assigns all the necessary env variables by prefix using BindEnvForStruct,
+// then it use viper.UnmarshalExact using the following custom hooks:
+//   - PublicKeyHook;
+//   - Uint32StrictHook;
+//   - Uint160Hook;
+//
+// Then it sets the found keys in the config and normalizes it.
+func Unmarshal(v *viper.Viper, cfg config, envPrefix string) error {
+	BindEnvForStruct(v, cfg, envPrefix)
+
+	if err := v.UnmarshalExact(cfg, viper.DecodeHook(
+		mapstructure.ComposeDecodeHookFunc(
+			mapstructure.StringToTimeDurationHookFunc(),
+			mapstructure.StringToSliceHookFunc(","),
+			PublicKeyHook(),
+			Uint32StrictHook(),
+			Uint160Hook(),
+		))); err != nil {
+		return fmt.Errorf("unable to decode config: %w", err)
+	}
+	for _, key := range v.AllKeys() {
+		cfg.Set(key)
+	}
+	cfg.Normalize()
+
+	return nil
+}
+
+// BindEnvForStruct goes through all fields of the structure,
+// searches for the corresponding env variables by tags, and calls viper.Set.
+func BindEnvForStruct(v *viper.Viper, s any, envPrefix string) {
+	bindFields(v, reflect.ValueOf(s), "", envPrefix)
+}
+
+func bindFields(v *viper.Viper, val reflect.Value, prefix string, envPrefix string) {
+	if val.Kind() == reflect.Ptr {
+		val = val.Elem()
+	}
+	if val.Kind() != reflect.Struct {
+		return
+	}
+
+	typ := val.Type()
+	for i := range val.NumField() {
+		field := typ.Field(i)
+		fieldVal := val.Field(i)
+
+		tag := field.Tag.Get("mapstructure")
+		if tag == "" || tag == "-" {
+			continue
+		}
+
+		tagParts := strings.Split(tag, ",")
+		fieldKey := tagParts[0]
+		isSquash := false
+		for _, part := range tagParts[1:] {
+			if part == "squash" {
+				isSquash = true
+				break
+			}
+		}
+
+		var viperKey string
+		if !isSquash && prefix != "" {
+			viperKey = prefix + "." + fieldKey
+		} else if !isSquash {
+			viperKey = fieldKey
+		} else {
+			viperKey = prefix
+		}
+
+		if !isSquash && fieldVal.Kind() != reflect.Struct && (fieldVal.Kind() != reflect.Ptr || fieldVal.Elem().Kind() != reflect.Struct) {
+			envKey := strings.ToUpper(envPrefix + "_" + strings.ReplaceAll(viperKey, ".", "_"))
+			envValue := os.Getenv(envKey)
+			if envValue != "" {
+				v.Set(viperKey, envValue)
+			}
+		}
+
+		if fieldVal.Kind() == reflect.Struct || (fieldVal.Kind() == reflect.Ptr && fieldVal.Elem().Kind() == reflect.Struct) {
+			bindFields(v, fieldVal, viperKey, envPrefix)
+		}
+	}
+}
+
+// PublicKeyHook returns a mapstructure decode hook func that converts a string to a keys.PublicKey.
+func PublicKeyHook() mapstructure.DecodeHookFunc {
+	return func(
+		f reflect.Type,
+		t reflect.Type,
+		data any,
+	) (any, error) {
+		if t == reflect.TypeOf(keys.PublicKey{}) {
+			str, ok := data.(string)
+			if !ok {
+				return nil, fmt.Errorf("invalid public key %v", data)
```

Comment:
when type is wrong, it doesnt matter what value it is
```suggestion
				return nil, fmt.Errorf("public key field type %T instead of %T", data, str)
```